### PR TITLE
Gemini integration

### PR DESCRIPTION
**Pull request by aleksbykov**
- integrate json format parse for result output file

### CODE REVIEW

File: test-cases/gemini/gemini-basic-3h.yaml [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
-gemini_cmd: "gemini -d -n 500000 -c 20 -p 100 -m mixed -f"
-gemini_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/gemini-latest'
+gemini_cmd: "gemini -d --duration 10800s -c 20 -p 100 -m mixed -f"
+gemini_url: 'http://downloads.scylladb.com.s3.amazonaws.com/gemini/0.9.2/gemini_0.9.2_Linux_x86_64.tar.gz'
 gemini_static_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/schema.json'
 
 db_type: mixed_scylla
-instance_type_db_oracle: 'i3.8xlarge'
-ami_id_db_oracle: 'ami-0081311880c437b65' # ScyllaDB 2.3.3 us-east-1 AMI
+instance_type_db_oracle: 'i3.large'
```
Comment (roydahan) [inline test-cases/gemini/gemini-basic-3h.yaml]: Why did you change this for the test_oracle?

File: test-cases/gemini/gemini-basic-3h.yaml [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
+# ami_id_db_oracle: 'ami-0081311880c437b65' # ScyllaDB 2.3.3 us-east-1 AMI
```
Comment (roydahan) [inline test-cases/gemini/gemini-basic-3h.yaml]: Why did you comment it?
What do we use for the test-oracle?

File: test-cases/gemini/gemini-basic-3h.yaml [hunk 3]
```diff
@@ -1,20 +1,20 @@
...
-failure_post_behavior: keep
+failure_post_behavior: destroy
 store_results_in_elasticsearch: False
 # gemini
 # cmd: gemini -d -n [NUM_OF_TEST_ITERATIONS] -c [NUM_OF_THREADS] -p [NUM_OF_PARTITION_KEYS_PER_THREAD] -m mixed -f
 # the below cmd runs about 3 hours
-gemini_cmd: "gemini -d -n 500000 -c 20 -p 100 -m mixed -f"
-gemini_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/gemini-latest'
+gemini_cmd: "gemini -d --duration 10800s -c 20 -p 100 -m mixed -f"
+gemini_url: 'http://downloads.scylladb.com.s3.amazonaws.com/gemini/0.9.2/gemini_0.9.2_Linux_x86_64.tar.gz'
```
Comment (roydahan) [inline test-cases/gemini/gemini-basic-3h.yaml]: We need to set it as a "east to replace param" like GEMINI_VERSION
So, the jenkins job will pass the version we want to use.

File: test-cases/gemini/gemini-basic-3h.yaml [hunk 4]
```diff
@@ -1,20 +1,20 @@
 test_duration: 500
 n_db_nodes: 1
 n_loaders: 1
-n_monitor_nodes: 1
+n_monitor_nodes: 0
 instance_type_db: 'i3.large'
 
 user_prefix: 'gemini-basic-3h'
-failure_post_behavior: keep
+failure_post_behavior: destroy
```
Comment (roydahan) [inline test-cases/gemini/gemini-basic-3h.yaml]: At least for the first runs we should run it with keep.
Then, we need to add a functionality that only in case of failure we keep.
@bentsi I think originally this was the idea behind this param, but somehow it became the general behaviour.

File: test-cases/gemini/gemini-basic-3h.yaml [hunk 5]
```diff
@@ -1,20 +1,20 @@
 test_duration: 500
 n_db_nodes: 1
```
Comment (roydahan) [inline test-cases/gemini/gemini-basic-3h.yaml]: This should be increased to 3

**Comment by roydahan** [comment]
Not enough just to store the results in Elastic Search.
We need to take the test results summary and send it in mail to RND.

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 1 above)
Comment (aleksbykov) [reply test-cases/gemini/gemini-basic-3h.yaml]: After talk with Pekka, he said that there is not different requirement to test_db_oracle instance. So i decided to have same instances for better test result correlation.

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 2 above)
Comment (aleksbykov) [reply test-cases/gemini/gemini-basic-3h.yaml]: This parameter is configured in Job configuration, and could be set upon run. In job i use the the 2.3.5 version of Scylla.

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 1 above)
Comment (roydahan) [reply test-cases/gemini/gemini-basic-3h.yaml]: There is a difference with how we run it.
We want that all the data to fit in memory so the the test_oracle won't involve disk operations.

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 2 above)
Comment (roydahan) [reply test-cases/gemini/gemini-basic-3h.yaml]: Even if this is the case, this is not the way to do it.
You could replacve the value with AMI_ID_TEST_ORACLE

However, I don't think this is something we want to replace every run or allow people to change it often so it won't cause confusions.
If you want to update to 2.3.5 just update here.

File: sdcm/cluster.py [hunk 6]
```diff
@@ -2474,9 +2475,11 @@ def node_setup(self, node, verbose=False, db_node_address=None, **kwargs):
...
             logger.warning('Gemini URLs should be defined to run the gemini tool')
         else:
+            gemini_tar = os.path.basename(gemini_url)
             install_gemini_script = dedent("""
                 cd $HOME
-                curl -L -o gemini {gemini_url}
+                curl -LO {gemini_url}
+                tar -xvf {gemini_tar}
                 chmod a+x gemini
                 curl -LO  {gemini_static_url}
```
Comment (bentsi) [inline sdcm/cluster.py]: what is `gemini_static_url` used for? consider rename

File: sdcm/cluster.py [hunk 7]
```diff
@@ -2766,15 +2769,29 @@ def _parse_cs_results(lines):
         return results
 
     @staticmethod
+    def _parse_gemini_summary_json(json_str):
+
```
Comment (bentsi) [inline sdcm/cluster.py]: remove redundant spaces

File: sdcm/cluster.py [hunk 8]
```diff
@@ -2766,15 +2769,29 @@ def _parse_cs_results(lines):
...
+        results = {'result': {}}
+        try:
+            results = json.loads(json_str)
+
+        except Exception as details:
+            logger.debug("Invalid json document {}".format(details))
```
Comment (bentsi) [inline sdcm/cluster.py]: why not return `{}` when handling exception?

File: sdcm/cluster.py [hunk 9]
```diff
@@ -2766,15 +2769,29 @@ def _parse_cs_results(lines):
...
+        return results.get('result')
+
     def _parse_gemini_summary(lines):
         results = {}
         enable_parse = False
 
         for line in lines:
             line.strip()
-            if line.startswith('Results:'):
+            if 'Results:' in line:
```
Comment (bentsi) [inline sdcm/cluster.py]: what if the line is empty?

File: sdcm/cluster.py (hunk 8 above)
Comment (bentsi) [inline sdcm/cluster.py]: logger.error

File: sdcm/tester.py [hunk 10]
```diff
@@ -790,9 +794,11 @@ def run_gemini(self, cmd, duration=None):
 
         timeout = self.get_duration(duration)
         time.sleep(30)
+        test_node = random.choice(self.db_cluster.nodes)
+        oracle_node = random.choice(self.cs_db_cluster.nodes)
         return self.loaders.run_gemini_thread(cmd, timeout, self.outputdir,
-                                              test_node=self.db_cluster.nodes[0].private_ip_address,
-                                              oracle_node=self.cs_db_cluster.nodes[0].private_ip_address)
+                                              test_node=test_node.private_ip_address,
```
Comment (bentsi) [inline sdcm/tester.py]: do you need `private_ip_address`? or you can use `ip_address`?

File: test-cases/gemini/gemini-basic-3h.yaml [hunk 11]
```diff
@@ -1,20 +1,20 @@
...
+failure_post_behavior: destroy
 store_results_in_elasticsearch: False
 # gemini
 # cmd: gemini -d -n [NUM_OF_TEST_ITERATIONS] -c [NUM_OF_THREADS] -p [NUM_OF_PARTITION_KEYS_PER_THREAD] -m mixed -f
 # the below cmd runs about 3 hours
-gemini_cmd: "gemini -d -n 500000 -c 20 -p 100 -m mixed -f"
-gemini_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/gemini-latest'
+gemini_cmd: "gemini -d --duration 10800s -c 20 -p 100 -m mixed -f"
+gemini_url: 'http://downloads.scylladb.com.s3.amazonaws.com/gemini/0.9.2/gemini_0.9.2_Linux_x86_64.tar.gz'
 gemini_static_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/schema.json'
```
Comment (bentsi) [inline test-cases/gemini/gemini-basic-3h.yaml]: `gemini_static_url` ->  `gemini_schema_url`

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 4 above)
Comment (bentsi) [reply test-cases/gemini/gemini-basic-3h.yaml]: agree, added a task in Trello

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 11 above)
Comment (bentsi) [reply test-cases/gemini/gemini-basic-3h.yaml]: add comment that it is currently not used and open issue

File: gemini_test.py [hunk 12]
```diff
@@ -38,4 +38,18 @@ def test_random_load(self):
         results = self.get_gemini_results(queue=test_queue)
         self.log.debug(results)
         assert results, "Gemini results are not found."
-        assert results[0]['read errors'] == 0, "Gemini read errors."
+        failed = False
+        for res in results:
+            if res['write_errors'] > 0:
```
[the remark this answers is not in the document]
Comment (aleksbykov) [reply gemini_test.py]: Agree. Fixed

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 11 above)
Comment (aleksbykov) [reply test-cases/gemini/gemini-basic-3h.yaml]: changed

File: sdcm/tester.py (hunk 10 above)
Comment (aleksbykov) [reply sdcm/tester.py]: fixed

File: sdcm/cluster.py (hunk 8 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: changed

File: sdcm/cluster.py (hunk 9 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: then enable_parse is not True.

File: sdcm/cluster.py (hunk 8 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: it will be returned anyway

File: sdcm/cluster.py (hunk 7 above)
Comment (aleksbykov) [reply sdcm/cluster.py]: done

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 4 above)
Comment (aleksbykov) [reply test-cases/gemini/gemini-basic-3h.yaml]: return keep

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 5 above)
Comment (aleksbykov) [reply test-cases/gemini/gemini-basic-3h.yaml]: increased

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 3 above)
Comment (aleksbykov) [reply test-cases/gemini/gemini-basic-3h.yaml]: Add new parameter Gemini version, which is used to build the gemini binary url to download

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 1 above)
Comment (aleksbykov) [reply test-cases/gemini/gemini-basic-3h.yaml]: return i3.8xlarge

**Comment by aleksbykov** [comment]
@roydahan @bentsi Resolved comments. The Storing and send mail functionality is in another PR #939

File: test-cases/gemini/gemini-basic-3h.yaml [hunk 13]
```diff
@@ -10,9 +10,9 @@ store_results_in_elasticsearch: False
 # gemini
 # cmd: gemini -d -n [NUM_OF_TEST_ITERATIONS] -c [NUM_OF_THREADS] -p [NUM_OF_PARTITION_KEYS_PER_THREAD] -m mixed -f
 # the below cmd runs about 3 hours
-gemini_cmd: "gemini -d -n 500000 -c 20 -p 100 -m mixed -f"
-gemini_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/gemini-latest'
-gemini_static_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/schema.json'
+gemini_cmd: "gemini -d --duration 120s -c 20 -p 100 -m mixed -f"
+gemini_version: '0.9.2'
```
Comment (roydahan) [inline test-cases/gemini/gemini-basic-3h.yaml]: According to this parameter we choose the gemini version to download Or is it just for annotation?
If the former, maybe it worth to link to latest or something so we won't need to manually replace the version every time.

Comment (bentsi) [reply test-cases/gemini/gemini-basic-3h.yaml]: it downloads gemini by version

Comment (bentsi) [reply test-cases/gemini/gemini-basic-3h.yaml]: let's keep version to track which version ran

File: test-cases/gemini/gemini-basic-3h.yaml [hunk 14]
```diff
@@ -1,7 +1,7 @@
 test_duration: 500
-n_db_nodes: 1
+n_db_nodes: 3
 n_loaders: 1
-n_monitor_nodes: 1
+n_monitor_nodes: 0
```
Comment (bentsi) [inline test-cases/gemini/gemini-basic-3h.yaml]: BTW: why number of monitors is 0? @roydahan

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 2 above)
Comment (aleksbykov) [reply test-cases/gemini/gemini-basic-3h.yaml]: done

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 13 above)
Comment (roydahan) [reply test-cases/gemini/gemini-basic-3h.yaml]: We can track the version we run by checking it and storing it in ealsticsearch.
How do you suggest to keep moving forward with Gemini versions?
We can ask them to transfer us the gemini version from their upstream job maybe, and replace it here if possible.

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 14 above)
Comment (roydahan) [reply test-cases/gemini/gemini-basic-3h.yaml]: Don't know, I think it's better to have it 1.

File: test-cases/gemini/gemini-basic-3h.yaml (hunk 13 above)
Comment (bentsi) [reply test-cases/gemini/gemini-basic-3h.yaml]: We can track by email and change manually when we want. But it should be a controlled process. Imagine that they released new version with a bug and you use always latest gemini job will fail.
image

Comment (roydahan) [reply test-cases/gemini/gemini-basic-3h.yaml]: As we discussed, we don't want to maintain it manually and we're ok if we find issues with Gemini that may cause failures.
We approves this for now and open a new task for tracking "latest" gemini and running with latest option.

File: test-cases/gemini/gemini-basic-3h.yaml [hunk 15]
```diff
@@ -10,9 +10,9 @@ store_results_in_elasticsearch: False
 # gemini
 # cmd: gemini -d -n [NUM_OF_TEST_ITERATIONS] -c [NUM_OF_THREADS] -p [NUM_OF_PARTITION_KEYS_PER_THREAD] -m mixed -f
 # the below cmd runs about 3 hours
-gemini_cmd: "gemini -d -n 500000 -c 20 -p 100 -m mixed -f"
-gemini_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/gemini-latest'
-gemini_static_url: 'https://s3.amazonaws.com/scylla-gemini/Binaries/schema.json'
+gemini_cmd: "gemini -d --duration 120s -c 20 -p 100 -m mixed -f"
```
Comment (roydahan) [inline test-cases/gemini/gemini-basic-3h.yaml]: Why is it 120s?
It should be 180m/3h (the job is 3h).
